File: 2461-maximum-sum-of-distinct-subarrays-with-length-k/2461-maximum-sum-of-distinct-subarrays-with-length-k.cpp
class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) {
        unordered_map<int,int> m;
        int i=0,j=0,n=nums.size();
        long long ans = 0;
        long long sum = 0;
        
        while(j<n){
            m[nums[j]]++;
            sum += nums[j];
            if(j-i+1 < k){
                j++;
            }
            else if(j-i+1 == k){
                if(m.size() == k){
                    ans = max(ans,sum);
                }
                m[nums[i]]--;
                if(m[nums[i]] == 0)m.erase(nums[i]);
                sum -= nums[i];
                i++,j++;
            }
        }
        return ans;
    }
};

File: 2461-maximum-sum-of-distinct-subarrays-with-length-k/2461-maximum-sum-of-distinct-subarrays-with-length-k.java
class Solution {
    public long maximumSubarraySum(int[] nums, int k) {
        HashMap<Integer,Integer> m = new HashMap<>();
        int n = nums.length;
        int i = 0;
        int j = 0;
        
        long ans = 0;
        long sum = 0;
        
        while(j<n){
            sum += nums[j];
            m.put(nums[j],m.getOrDefault(nums[j],0)+1);
            if(j-i+1 < k){
                j++;
            }
            else if(j-i+1 == k){
                if(m.size() == k){
                    ans = Math.max(ans,sum);
                }
                sum -= nums[i];
                m.put(nums[i],m.getOrDefault(nums[i],1)-1);
                if(m.get(nums[i]) == 0)
                    m.remove(nums[i]);
                i++;
                j++;
            }
        }
        return ans;
    }
}